File: apps/macos/Kelpie/Account/UOATransport.swift
import Foundation

/// Credentials never enter renderer cookies, disk caches, redirect targets or logs.
final class UOATransport: NSObject, URLSessionTaskDelegate {
    static let origin = "https://authentication.unlikeotherai.com"
    private lazy var session: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.httpCookieStorage = nil
        configuration.urlCache = nil
        configuration.timeoutIntervalForRequest = 30
        return URLSession(configuration: configuration, delegate: self, delegateQueue: nil)
    }()

    struct Response {
        let data: Data
        let version: String?
    }

    struct Failure: LocalizedError {
        let status: Int
        var errorDescription: String? {
            switch status {
            case 401: return "Your UOA session has expired. Sign in again."
            case 403: return "Your UOA account did not grant access to favourites."
            case 409: return "Favourites changed on another device. Please try again."
            default: return "UOA could not complete this request (\(status)). Please try again."
            }
        }
    }

    func request(_ path: String, method: String = "GET", token: String? = nil, body: Data? = nil, version: String? = nil) async throws -> Response {
        guard path.hasPrefix("/oauth/"), !path.contains(".."),
              let url = URL(string: Self.origin + path), url.host == "authentication.unlikeotherai.com" else {
            throw Failure(status: 400)
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let version { request.setValue(version, forHTTPHeaderField: "If-Match") }
        let (data, response) = try await session.data(for: request)
        guard let response = response as? HTTPURLResponse else { throw Failure(status: 0) }
        guard (200..<300).contains(response.statusCode) else { throw Failure(status: response.statusCode) }
        return Response(data: data, version: response.value(forHTTPHeaderField: "ETag"))
    }

    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/account/UOATransport.kt
package com.kelpie.browser.account

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import java.net.HttpURLConnection
import java.net.URL

/** No cookies, redirect following, disk caching or logging of identity requests. */
class UOATransport {
    data class Response(
        val data: ByteArray,
        val version: String?,
    ) {
        fun json(): JsonObject = Json.parseToJsonElement(data.toString(Charsets.UTF_8)).jsonObject
    }

    class Failure(
        val status: Int,
    ) : Exception(
            when (status) {
                401 -> "Your session expired. Please log in again."
                403 -> "Access to favourites was not granted."
                409 -> "Favourites changed on another device. Please try again."
                else -> "Could not complete the request. Please try again."
            },
        )

    suspend fun request(
        path: String,
        method: String = "GET",
        token: String? = null,
        body: String? = null,
        version: String? = null,
    ): Response =
        withContext(Dispatchers.IO) {
            require(path.startsWith("/oauth/") && !path.contains("..") && !path.contains('?'))
            val connection = URL(ORIGIN + path).openConnection() as HttpURLConnection
            try {
                connection.requestMethod = method
                connection.instanceFollowRedirects = false
                connection.useCaches = false
                connection.connectTimeout = 30000
                connection.readTimeout = 30000
                connection.setRequestProperty("Content-Type", "application/json")
                if (token != null) connection.setRequestProperty("Authorization", "Bearer $token")
                if (version != null) connection.setRequestProperty("If-Match", version)
                if (body != null) {
                    connection.doOutput = true
                    connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
                }
                if (connection.responseCode !in 200..299) throw Failure(connection.responseCode)
                val bytes =
                    connection.inputStream.use { stream ->
                        val output = java.io.ByteArrayOutputStream()
                        val buffer = ByteArray(8192)
                        while (true) {
                            val count = stream.read(buffer)
                            if (count < 0) break
                            require(output.size() + count <= 5 * 1024 * 1024)
                            output.write(buffer, 0, count)
                        }
                        output.toByteArray()
                    }
                require(bytes.size <= 5 * 1024 * 1024)
                Response(bytes, connection.getHeaderField("ETag"))
            } finally {
                connection.disconnect()
            }
        }

    companion object {
        const val ORIGIN = "https://authentication.unlikeotherai.com"
    }
}
